// erela.js/dist/lib/utils/Store.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Store extends Map {
    constructor(iterable) {
        // @ts-ignore
        super(iterable);
    }
    get(key) {
        return super.get(key);
    }
    set(key, value) {
        return super.set(key, value);
    }
    find(fn) {
        if (typeof fn !== "function") {
            throw new Error("First argument must be a function.");
        }
        for (const [key, val] of this) {
            if (fn(val, key, this)) {
                return val;
            }
        }
        return null;
    }
    first(count) {
        if (count === undefined) {
            return this.values().next().value;
        }
        if (typeof count !== "number") {
            throw new TypeError("The count must be a number.");
        }
        if (!Number.isInteger(count) || count < 1) {
            throw new RangeError("The count must be an integer greater than 0.");
        }
        count = Math.min(this.size, count);
        const arr = new Array(count);
        const iter = this.values();
        for (let i = 0; i < count; i++) {
            arr[i] = iter.next().value;
        }
        return arr;
    }
    filter(fn) {
        const results = new Store();
        for (const [key, val] of this) {
            if (fn(val, key, this)) {
                results.set(key, val);
            }
        }
        return results;
    }
    map(fn) {
        const arr = new Array(this.size);
        let i = 0;
        for (const [key, val] of this) {
            arr[i++] = fn(val, key, this);
        }
        return arr;
    }
    some(fn) {
        for (const [key, val] of this) {
            if (fn(val, key, this)) {
                return true;
            }
        }
        return false;
    }
    concat(...collections) {
        const newColl = new Store(this);
        for (const coll of collections) {
            for (const [key, val] of coll) {
                newColl.set(key, val);
            }
        }
        return newColl;
    }
    sort(compareFunction = (x, y) => +(x > y) || +(x === y) - 1) {
        // @ts-ignore
        return new Store([...this.entries()].sort((a, b) => compareFunction(a[1], b[1], a[0], b[0])));
    }
}
exports.default = Store;
