// erela.js/dist/lib/ErelaClient.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// tslint:disable-next-line: no-var-requires
const axios_1 = __importDefault(require("axios"));
const events_1 = require("events");
const Node_1 = require("./entities/Node");
const Player_1 = require("./entities/Player");
const Queue_1 = require("./entities/Queue");
const SearchResult_1 = require("./entities/SearchResult");
const Track_1 = require("./entities/Track");
const PlayerStore_1 = __importDefault(require("./stores/PlayerStore"));
const NodeStore_1 = __importDefault(require("./stores/NodeStore"));
const libraries_1 = __importDefault(require("./libraries"));
const lodash_1 = __importDefault(require("lodash"));
const defaultOptions = {
    shardCount: 1,
    player: Player_1.Player,
    node: Node_1.Node,
    track: Track_1.Track,
    queue: Queue_1.Queue,
};
const isInstalled = (name) => {
    try {
        require(name);
        return true;
    }
    catch (e) {
        return false;
    }
};
/**
 * The Erela class.
 * @noInheritDoc
 */
class ErelaClient extends events_1.EventEmitter {
    /**
     * Emitted when a player is created.
     * @event ErelaClient#playerCreate
     * @param {Player} player - The created player.
     */
    /**
     * Emitted when a player is destroyed.
     * @event ErelaClient#playerDestroy
     * @param {Player} player - The destroyed player.
     */
    /**
     * Emitted when a player is moved to a new channel.
     * @event ErelaClient#playerMove
     * @param {Player} player - The moved player.
     * @param {any} oldChannel - The old voice channel.
     * @param {any} newChannel - The new voice channel.
     */
    /**
     * Emitted when a track is started.
     * @event ErelaClient#trackStart
     * @param {Player} player - The player that has the track.
     * @param {Track} track - The track that started.
     */
    /**
     * Emitted when a track ends.
     * @event ErelaClient#trackEnd
     * @param {Player} player - The player that has the track.
     * @param {Track} track - The track that ended.
     */
    /**
     * Emitted when a track gets stuck during playback.
     * @event ErelaClient#trackStuck
     * @param {Player} player - The player that has the track.
     * @param {Track} track - The track that ended.
     * @param {*} message - The message for the event.
     */
    /**
     * Emitted when a track errors during playback.
     * @event ErelaClient#trackStuck
     * @param {Player} player - The player that has the track.
     * @param {Track} track - The track that ended.
     * @param {*} message - The message for the event.
     */
    /**
     * Emitted when a queue ends.
     * @event ErelaClient#queueEnd
     * @param {Player} player - The player who's queue has ended.
     */
    /**
     * Emitted when a player voice channel connected is closed.
     * @event ErelaClient#socketClosed
     * @param {Player} player - The player.
     * @param {any} message - The message.
     */
    /**
     * Emitted when a node is created.
     * @event ErelaClient#nodeCreate
     * @param {Node} node - The created node.
     */
    /**
     * Emitted when a node connects.
     * @event ErelaClient#nodeConnect
     * @param {Node} node - The node that connected.
     */
    /**
     * Emitted when a node reconnects.
     * @event ErelaClient#nodeReconnect
     * @param {Node} node - The node that reconnected.
     */
    /**
     * Emitted when a node encounters an error.
     * @event ErelaClient#nodeError
     * @param {Node} node - The node.
     * @param {Error} error - The error.
     */
    /**
     * Emitted when a node disconnects abnormally.
     * @event ErelaClient#nodeDisconnect
     * @param {Node} node - The node.
     * @param {Error} error - The error.
     */
    /**
     * Creates an instance of ErelaClient.
     * @param {*} client - The Discord client.
     * @param {Array<INodeOptions>} nodes - The nodes to use.
     * @param {IErelaOptions} [options=defaultOptions] - Options for the client.
     */
    constructor(client, nodes, options) {
        super();
        /**
         * The shard count.
         */
        this.shardCount = 1;
        this.voiceState = new Map();
        // @ts-ignore
        if (process._erela_client_defined) {
            return;
        }
        let _nodes = [];
        let _options = defaultOptions;
        const isObject = (obj) => typeof obj !== "undefined" && JSON.stringify(obj)[0] === "{";
        const isClass = (obj) => typeof obj !== "undefined" && obj.constructor.toString().includes("class");
        if (lodash_1.default.isObject(client) && lodash_1.default.isArray(nodes) && !lodash_1.default.isUndefined(options)) { // client, nodes, options
            this.client = client;
            _nodes = nodes;
            _options = options;
        }
        else if (lodash_1.default.isArray(client) && isObject(nodes) && lodash_1.default.isUndefined(options)) { // nodes, options
            _nodes = client;
            _options = nodes;
        }
        else if (lodash_1.default.isObject(client) && lodash_1.default.isArray(nodes) && lodash_1.default.isUndefined(options)) { // client, nodes
            this.client = client;
            _nodes = nodes;
        }
        else if (lodash_1.default.isObject(client) && isObject(nodes) && lodash_1.default.isUndefined(options)) { // client, options
            this.client = client;
            _options = nodes;
        }
        else if (lodash_1.default.isArray(client) && lodash_1.default.isUndefined(nodes) && lodash_1.default.isUndefined(options)) { // nodes
            _nodes = client;
        }
        else if (isClass(client) && lodash_1.default.isUndefined(nodes) && lodash_1.default.isUndefined(options)) { // client
            this.client = client;
        }
        else if (isObject(client) && lodash_1.default.isUndefined(nodes) && lodash_1.default.isUndefined(options)) { // options
            _options = client;
        }
        this.userId = isClass(this.client) && !lodash_1.default.isNull(this.client.user) ? this.client.user.id : _options.userId;
        if (!this.userId) {
            throw new RangeError("new ErelaClient() No user ID supplied.");
        }
        const found = Object.keys(libraries_1.default).find(isInstalled);
        this.library = libraries_1.default[_options.library] ?
            libraries_1.default[_options.library] : found ?
            libraries_1.default[found] : undefined;
        this.shardCount = _options.shardCount || 1;
        this.node = _options.node || Node_1.Node;
        this.player = _options.player || Player_1.Player;
        this.queue = _options.queue || Queue_1.Queue;
        this.track = _options.track || Track_1.Track;
        this.players = new PlayerStore_1.default(this);
        this.nodes = new NodeStore_1.default(this, _nodes);
        if (isClass(client)) {
            client.on(this.library.ws.string, this.updateVoiceState.bind(this));
        }
    }
    /**
     * Sends voice data to the Lavalink server. Only use this when you're using a unsupported library.
     * @param {*} data
     */
    updateVoiceState(data) {
        if (!data || !["VOICE_SERVER_UPDATE", "VOICE_STATE_UPDATE"].includes(data.t || "")) {
            return;
        }
        const player = this.players.get(data.d.guild_id);
        if (!player) {
            return;
        }
        const state = this.voiceState.get(data.d.guild_id) || {};
        if (data.t === "VOICE_SERVER_UPDATE") {
            state.op = "voiceUpdate";
            state.guildId = data.d.guild_id;
            state.event = data.d;
        }
        else {
            if (data.d.user_id !== this.userId) {
                return;
            }
            state.sessionId = data.d.session_id;
            const channel = player.voiceChannel.id || player.voiceChannel;
            if (channel !== data.d.channel_id) {
                const found = !this.library ? data.d.channel_id :
                    this.library.findChannel(this.client, data.d.channel_id);
                const currentChannel = player.voiceChannel.id ? player.voiceChannel : player.voiceChannel;
                const newChannel = player.voiceChannel.id ? found : found.id;
                this.emit("playerMove", player, currentChannel, newChannel);
                player.voiceChannel = newChannel;
            }
        }
        this.voiceState.set(data.d.guild_id, state);
        const template = JSON.stringify(["op", "guildId", "sessionId", "event"].sort());
        if (JSON.stringify(Object.keys(state).sort()) === template) {
            player.node.send(state);
            this.voiceState.set(data.d.guild_id, {});
        }
    }
    /**
     * Searches YouTube with the query.
     * @param {(string|IQuery)} query - The query to search against.
     * @param {any} user - The user who requested the tracks.
     * @returns {Promise<SearchResult>} - The search result.
     */
    search(query, user) {
        return new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            const node = this.nodes.leastUsed.first();
            if (!node) {
                throw new Error("ErelaClient#search() No available nodes.");
            }
            const source = { soundcloud: "sc" }[query.source] || "yt";
            let search = query.query || query;
            if (!/^https?:\/\//.test(search)) {
                search = `${source}search:${search}`;
            }
            const url = `http://${node.options.host}:${node.options.port}/loadtracks`;
            const res = yield axios_1.default.get(url, {
                headers: { Authorization: node.options.password },
                params: { identifier: search },
            }).catch((err) => {
                return reject(err);
            });
            node.calls++;
            if (!res || !res.data || !res.data.tracks) {
                return reject(new Error("No data returned from query."));
            }
            if (!res.data.tracks[0]) {
                return reject(new Error("No tracks were found."));
            }
            const result = new SearchResult_1.SearchResult(res.data, this.track, user);
            return resolve(result);
        }));
    }
    /**
     * Sends data to Discord via WebSocket, only available when using a supported library.
     * @param {*} data
     */
    sendWS(data) {
        if (!this.client) {
            return;
        }
        const guild = this.library.findGuild(this.client, data.d.guild_id);
        if (guild && this.library.isSharded(this.client)) {
            this.library.sendShardWS(guild, data);
        }
        else if (guild) {
            this.library.sendWS(this.client, data);
        }
    }
}
exports.ErelaClient = ErelaClient;
